`org.openrewrite.java.migrate.util.UseMapOf` fails on double brace init for non-string types

## What is the smallest, simplest way to reproduce the problem?
```java
class A {
   import static foo.BLAH;  // Some string 
    void foo() {
                  new HashMap<>() {{
                    put(BLAH, "foo");
                }}
    }
}
```

## What did you expect to see?
```java
import static foo.BLAH;  // Some string 
Map.of(BLAH, "foo")
```

The problem is that https://github.com/openrewrite/rewrite-migrate-java/blob/main/src/main/java/org/openrewrite/java/migrate/util/UseMapOf.java#L66 is using untyped which fails for non-trivial types

## What did you see instead?


## What is the full stack trace of any errors you encountered?
```
Caused by: java.lang.RuntimeException: Error while visiting tst/com/amazon/javelinprimingservice/creativemetadata/AmentumAdaptorTest.java: java.lang.IllegalArgumentException: Template parameter 1 cannot be used in an untyped template substitution. Instead of "#{}", indicate the template parameter's type with "#{any(java.lang.String)}".
  org.openrewrite.java.internal.template.Substitutions.substituteUntyped(Substitutions.java:206)
  org.openrewrite.java.internal.template.Substitutions.lambda$substitute$0(Substitutions.java:82)
  org.openrewrite.internal.PropertyPlaceholderHelper.parseStringValue(PropertyPlaceholderHelper.java:98)
  org.openrewrite.internal.PropertyPlaceholderHelper.replacePlaceholders(PropertyPlaceholderHelper.java:72)
  org.openrewrite.java.internal.template.Substitutions.substitute(Substitutions.java:54)
  org.openrewrite.java.JavaTemplate.apply(JavaTemplate.java:109)
  org.openrewrite.java.migrate.util.UseMapOf$1.visitNewClass(UseMapOf.java:79)
  org.openrewrite.java.migrate.util.UseMapOf$1.visitNewClass(UseMapOf.java:51)```
```

--- a/src/main/java/org/openrewrite/java/migrate/util/UseMapOf.java
+++ b/src/main/java/org/openrewrite/java/migrate/util/UseMapOf.java
@@ -48,40 +48,41 @@ public class UseMapOf extends Recipe {
 
     @Override
     public TreeVisitor<?, ExecutionContext> getVisitor() {
-        return Preconditions.check(Preconditions.and(new UsesJavaVersion<>(10), new UsesMethod<>(NEW_HASH_MAP)), new JavaVisitor<ExecutionContext>() {
-            @Override
-            public J visitNewClass(J.NewClass newClass, ExecutionContext ctx) {
-                J.NewClass n = (J.NewClass) super.visitNewClass(newClass, ctx);
-                J.Block body = n.getBody();
-                if (NEW_HASH_MAP.matches(n) && body != null) {
-                    if (body.getStatements().size() == 1) {
-                        Statement statement = body.getStatements().get(0);
-                        if (statement instanceof J.Block) {
-                            List<Expression> args = new ArrayList<>();
-                            StringJoiner mapOf = new StringJoiner(", ", "Map.of(", ")");
-                            for (Statement stat : ((J.Block) statement).getStatements()) {
-                                if (stat instanceof J.MethodInvocation && MAP_PUT.matches((Expression) stat)) {
-                                    J.MethodInvocation put = (J.MethodInvocation) stat;
-                                    args.addAll(put.getArguments());
-                                    mapOf.add("#{}");
-                                    mapOf.add("#{}");
-                                } else {
-                                    return n;
-                                }
-                            }
+        return Preconditions.check(Preconditions.and(new UsesJavaVersion<>(10),
+            new UsesMethod<>(NEW_HASH_MAP)), new UseMapOfVisitor());
+    }
 
-                            maybeRemoveImport("java.util.HashMap");
-                            maybeAddImport("java.util.Map");
-                            return JavaTemplate.builder(mapOf.toString())
-                                    .contextSensitive()
-                                    .imports("java.util.Map")
-                                    .build()
-                                    .apply(updateCursor(n), n.getCoordinates().replace(), args.toArray());
+    private static class UseMapOfVisitor extends JavaVisitor<ExecutionContext> {
+        @Override
+        public J visitNewClass(J.NewClass newClass, ExecutionContext ctx) {
+            J.NewClass n = (J.NewClass) super.visitNewClass(newClass, ctx);
+            J.Block body = n.getBody();
+            if (NEW_HASH_MAP.matches(n) && body != null && body.getStatements().size() == 1) {
+                Statement statement = body.getStatements().get(0);
+                if (statement instanceof J.Block) {
+                    List<Expression> args = new ArrayList<>();
+                    StringJoiner mapOf = new StringJoiner(", ", "Map.of(", ")");
+                    for (Statement stat : ((J.Block) statement).getStatements()) {
+                        if (!(stat instanceof J.MethodInvocation) || !MAP_PUT.matches((Expression) stat)) {
+                            return n;
                         }
+                        J.MethodInvocation put = (J.MethodInvocation) stat;
+                        args.addAll(put.getArguments());
+                        mapOf.add("#{any()}");
+                        mapOf.add("#{any()}");
                     }
+
+                    maybeRemoveImport("java.util.HashMap");
+                    maybeAddImport("java.util.Map");
+                    return JavaTemplate.builder(mapOf.toString())
+                        .contextSensitive()
+                        .imports("java.util.Map")
+                        .build()
+                        .apply(updateCursor(n), n.getCoordinates().replace(), args.toArray());
                 }
-                return n;
             }
-        });
+
+            return n;
+        }
     }
 }

--- a/src/test/java/org/openrewrite/java/migrate/util/UseMapOfTest.java
+++ b/src/test/java/org/openrewrite/java/migrate/util/UseMapOfTest.java
@@ -17,6 +17,7 @@ package org.openrewrite.java.migrate.util;
 
 import org.junit.jupiter.api.Test;
 import org.openrewrite.DocumentExample;
+import org.openrewrite.Issue;
 import org.openrewrite.test.RecipeSpec;
 import org.openrewrite.test.RewriteTest;
 
@@ -53,6 +54,41 @@ class UseMapOfTest implements RewriteTest {
 
               class Test {
                   Map<String, String> m = Map.of("stru", "menta", "mod", "erne");
+              }
+              """
+          )
+        );
+    }
+
+    @Issue("https://github.com/openrewrite/rewrite-migrate-java/issues/566")
+    @Test
+    void changeDoubleBraceInitForNonStringTypes() {
+        //language=java
+        rewriteRun(
+          java(
+            """
+              import java.util.HashMap;
+              import java.util.Map;
+
+              class Test {
+              private static final String BLAH ="ss";
+              
+              void foo() {
+                        new HashMap<>() {{
+                          put(BLAH, "foo");
+                      }};
+                  }
+              }
+              """,
+            """
+              import java.util.Map;
+
+              class Test {
+              private static final String BLAH ="ss";
+              
+              void foo() {
+                  Map.of(BLAH, "foo");
+                  }
               }
               """
           )

